### PR TITLE
a simple, short refactoring of lambda/Matching.do_compile_matching

### PR DESCRIPTION
**Pull request by gasche**
This refactoring-only PR shares about 5 common arguments between 8 different function calls inside the pattern-matching compiler. This makes it sensibly easier to tweak the arguments passed to pattern-matching compilation (in the context of The Pattern-Matching Bug). This commit was extracted from #13154. It is short and requires no domain-specific knowledge, so it should be easy to review.

### CODE REVIEW

**Comment by redianthus** [comment]
It looks correct to me and this is a clear improvement in terms of readability.

**Comment by gasche** [comment]
This needs the approval of a maintainer to move forward.

File: lambda/matching.ml
Comment (yallop) [inline lambda/matching.ml:3716]:
```suggestion
```

This local `open` isn't needed here and made the change slightly harder to review.

**Review by yallop — APPROVED** [review]

File: lambda/matching.ml
Comment (gasche) [reply lambda/matching.ml:3716]: I think that the person who wrote this `open` (probably me?) wanted to avoid relying on type-based disambiguation of constructors that are not in the lexical scope, a feature I still have mixed feelings about. I'll go with the easiest option of changing nothing at all here.

**Comment by gasche** [comment]
Thanks @zapashcanon and @yallop!

**Pull request merged**
